File: 004 HackerRank.py
t = int(input().strip())
for a0 in range(t):
    limit = int(input().strip())
    
    maxproduct,maxy,maxx = 0,0,0

    for x in range(990,100,-11):            # x divides by 11
            for y in range(999,100,-1):
                    product = x*y
                    if product < maxproduct and product< limit:
                            break
                    if product == int(str(product)[::-1]) and product< limit :
                            maxproduct,maxx,maxy = product,x,y
                            break
    # print("x =" , maxx)
    # print("y =" , maxy)
    print(maxproduct)

File: 004 HackerRank.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        for (int a0 = 0; a0 < t; a0++) {
            int limit = scanner.nextInt();
            int maxProduct = 0;
            int maxX = 0;
            int maxY = 0;

            for (int x = 990; x >= 100; x -= 11) {  // x is divisible by 11
                for (int y = 999; y >= 100; y--) {
                    int product = x * y;
                    if (product < maxProduct && product < limit) {
                        break;  // No need to check smaller y's
                    }
                    if (isPalindrome(product) && product < limit) {
                        if (product > maxProduct) {
                            maxProduct = product;
                            maxX = x;
                            maxY = y;
                        }
                        break;  // Found the largest palindrome for this x
                    }
                }
            }
            System.out.println(maxProduct);
        }
        scanner.close();
    }

    private static boolean isPalindrome(int num) {
        String s = Integer.toString(num);
        return s.equals(new StringBuilder(s).reverse().toString());
    }
}